src/autolabel/few_shot/label_diversity_example_selector.py

```diff
@@ -0,0 +1,153 @@
+"""Example selector that selects examples based on SemanticSimilarity."""
+from __future__ import annotations
+
+from itertools import groupby
+from operator import itemgetter
+from typing import Any, Dict, List, Optional, Type
+
+from langchain.embeddings.base import Embeddings
+from langchain.prompts.example_selector.base import BaseExampleSelector
+from langchain.vectorstores.base import VectorStore
+from pydantic import BaseModel, Extra
+
+
+def sorted_values(values: Dict[str, str]) -> List[Any]:
+    """Return a list of values in dict sorted by key."""
+    return [values[val] for val in sorted(values)]
+
+
+class LabelDiversityRandomExampleSelector(BaseExampleSelector, BaseModel):
+    """Example selector that selects examples based on SemanticSimilarity."""
+
+    examples: List[dict]
+    """A list of the examples that the prompt template expects."""
+    k: int = 1
+    """Number of examples to select per label."""
+    label_key: str
+    """Name of the label column/key."""
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        extra = Extra.forbid
+        arbitrary_types_allowed = True
+
+    def add_example(self, example: Dict[str, str]) -> None:
+        self.examples.append(example)
+
+    def select_examples(self, input_variables: Dict[str, str]) -> List[dict]:
+        selected_examples = []
+        sorted_examples = sorted(self.examples, key=itemgetter(self.label_key))
+        for label, label_examples in groupby(
+            sorted_examples, key=itemgetter(self.label_key)
+        ):
+            label_examples_list = list(label_examples)
+            selected_examples.extend(label_examples_list[: self.k])
+        return selected_examples
+
+    @classmethod
+    def from_examples(
+        cls,
+        examples: List[dict],
+        label_key: str,
+        k: int = 4,
+    ) -> LabelDiversityRandomExampleSelector:
+        """Create label diversity example selector using example list and embeddings.
+
+        Args:
+            examples: List of examples to use in the prompt.
+            k: Number of examples to select per label
+            label_key: Determines which variable corresponds to the example's label
+
+        Returns:
+            The ExampleSelector instantiated
+        """
+        return cls(k=k, examples=examples, label_key=label_key)
+
+
+class LabelDiversitySimilarityExampleSelector(BaseExampleSelector, BaseModel):
+    """ExampleSelector that selects examples based on label diversity, while choosing the most similar examples for each label"""
+
+    vectorstore: VectorStore
+    """VectorStore than contains information about examples."""
+    k: int = 4
```

Comment:
why are we setting the num examples per label default value different here (k = 4 vs k = 1) in `LabelDiversityRandomExampleSelector` ?